File: frontend/src/lib/i18n.ts
/**
 * The console in English or Spanish.
 *
 * Two kinds of text share this screen and they are translated in different places.
 *
 *   The engine's sentences - headlines, health bands, recommended actions, receipts -
 *   are composed from live values by si_core and rendered by a catalogue in si_api.
 *   The client never translates those; it asks for a language and displays what comes
 *   back. That is why every API call carries `lang`.
 *
 *   The console's own chrome - labels, column headings, legends, tooltips - is this
 *   file. It never contains a measurement, so it can be a plain dictionary.
 *
 * The split matters: a number the engine measured must not pass through a translation
 * table on its way to the screen, or the demo can no longer claim its console and its
 * scorecard agree.
 */

export const LANGS = ['en', 'es'] as const
export type Lang = (typeof LANGS)[number]

const STORAGE_KEY = 'movistar-si.lang'

/** The language to open in: the operator's last choice, else the browser's, else EN. */
export function initialLang(): Lang {
  try {
    const saved = window.localStorage.getItem(STORAGE_KEY)
    if (saved && (LANGS as readonly string[]).includes(saved)) return saved as Lang
  } catch {
    /* private browsing or storage disabled; fall through to the browser's preference */
  }
  const preferred = (window.navigator.languages ?? [window.navigator.language]).find((tag) =>
    (LANGS as readonly string[]).includes(tag?.split('-')[0]),
  )
  return (preferred?.split('-')[0] as Lang) ?? 'en'
}

export function rememberLang(lang: Lang): void {
  try {
    window.localStorage.setItem(STORAGE_KEY, lang)
  } catch {
    /* nothing to do: the choice simply will not survive a reload */
  }
}

/**
 * The console's own words. Keys read as what the string is for, not as the English of
 * it, so changing the English wording never means renaming a key.
 */
const STRINGS = {
  // ----- header and run controls -----
  'header.scale': ['Scale', 'Escala'],
  'header.scale.tiny': ['Tiny', 'Reducida'],
  'header.scale.full': ['Full', 'Completa'],
  'header.language': ['Language', 'Idioma'],
  'header.subtitle': ['Customer Experience Operations Center', 'Centro de Operaciones de Experiencia de Cliente'],
  'header.synthetic': ['Synthetic demo', 'Demo sintética'],
  'header.engine': ['SI engine', 'Motor SI'],
  'header.engine.running': ['running', 'ejecutando'],
  'header.engine.live': ['live', 'activo'],
  'header.scale.tiny.tip': [
    'A small network, a short run — the whole story in a second.',
    'Una red pequeña, una ejecución corta: toda la historia en un segundo.',
  ],
  'header.scale.full.tip': [
    'The full six-region network. Slower to compute.',
    'La red completa de seis regiones. Más lenta de calcular.',
  ],
  'alerts.formingNotResolved': ['Shape forming — not yet resolved', 'Firma formándose: aún sin resolver'],

  // ----- the closed vocabularies, as they read on a chip -----
  // These live here rather than in format.ts because they are wording, and format.ts
  // deliberately holds no language. The engine's own layer/shape codes are unchanged.
  'layerLabel.home': ['Home / Wi-Fi', 'Hogar / Wi-Fi'],
  'layerLabel.access': ['Access / aggregation', 'Acceso / agregación'],
  'layerLabel.core': ['Core / transport', 'Core / transporte'],
  'layerLabel.content': ['Content / CDN', 'Contenido / CDN'],
  'layerLabel.unresolved': ['Unresolved', 'Sin resolver'],
  'shapeLabel.single': ['Single (isolated home)', 'Individual (hogar aislado)'],
  'shapeLabel.cluster': ['Cluster (behind one node)', 'Clúster (tras un mismo nodo)'],
  'shapeLabel.path': ['Path (along a route)', 'Camino (a lo largo de una ruta)'],
  'shapeLabel.source': ['Source (one origin, many homes)', 'Fuente (un origen, muchos hogares)'],
  'shapeLabel.none': ['No shape', 'Sin firma'],

  // ----- the shape-graph legend -----
  'graph.impaired': ['impaired home', 'hogar afectado'],
  'graph.drifting': ['just starting to drift', 'empezando a desviarse'],
  'graph.healthyPeer': ['healthy peer', 'vecino sano'],
  'graph.namedElement': ['the named element', 'el elemento identificado'],

  // ----- tabs -----
  'tab.analyst': ['Analyst', 'Analista'],
  'tab.map': ['Map', 'Mapa'],
  'tab.management': ['Management', 'Dirección'],

  // ----- the alert rail -----
  'alerts.active': ['Active alerts', 'Alertas activas'],
  'alerts.homes': ['homes', 'hogares'],
  'alerts.home': ['home', 'hogar'],
  'alerts.competenceBoundary': ['Competence boundary', 'Límite de competencia'],
  'alerts.abstainedAt': ['abstained at interval', 'abstención en el intervalo'],

  // ----- honest instruments -----
  'honest.title': ['Honest instruments', 'Instrumentos de honestidad'],
  'honest.subtitle': [
    'Measured over the whole run, against ground truth the engine never sees.',
    'Medido sobre toda la ejecución, frente a una verdad de referencia que el motor nunca ve.',
  ],
  'honest.decoysFired': ['Decoys that fired', 'Señuelos que dispararon'],
  'honest.falsePositiveRate': ['False-positive rate', 'Tasa de falsos positivos'],
  'honest.quietIntervals': ['Quiet intervals watched', 'Intervalos en calma vigilados'],

  // ----- the verdict panel -----
  'verdict.element': ['element', 'elemento'],
  'verdict.layer': ['layer', 'capa'],
  'verdict.shape': ['shape', 'firma'],
  'verdict.region': ['region', 'región'],
  'verdict.confidence': ['confidence', 'confianza'],
  'verdict.caught': ['caught', 'detectado'],
  'verdict.beforeSurfacing': ['before it would surface', 'antes de que aflorara'],
  'verdict.theCall': ['The call', 'El veredicto'],
  'verdict.shapeOnGraph': ['The shape on the graph', 'La firma sobre el grafo'],
  'verdict.detectionNarrative': ['Detection narrative', 'Relato de la detección'],
  'verdict.namedAt': ['Named at interval', 'Identificado en el intervalo'],

  // ----- measured against ground truth -----
  'measured.title': ['Measured against ground truth', 'Medido frente a la verdad de referencia'],
  'measured.homesAffected': ['Homes affected', 'Hogares afectados'],
  'measured.leadTime': ['Lead time', 'Antelación'],
  'measured.localization': ['Localization', 'Localización'],
  'measured.layerAttribution': ['Layer attribution', 'Atribución de capa'],

  // ----- the receipt -----
  'receipt.title': ['Certified-decision receipt', 'Recibo de decisión certificada'],
  'receipt.claim': ['Claim', 'Afirmación'],
  'receipt.evidence': ['Evidence', 'Evidencia'],
  'receipt.check': ['Check', 'Contraste'],
  'receipt.provenance': ['Provenance', 'Procedencia'],
  'receipt.confidence': ['Confidence', 'Confianza'],

  // ----- recommended action -----
  'action.title': ['Recommended action', 'Acción recomendada'],
  'action.footer': [
    'These are the actions the verdict licenses. Nothing here dispatches to a real network.',
    'Estas son las acciones que autoriza el veredicto. Nada de esto se envía a una red real.',
  ],


  // ----- the transport bar -----
  'transport.interval': ['Interval', 'Intervalo'],
  'transport.play': ['Play the run', 'Reproducir la ejecución'],
  'transport.pause': ['Pause', 'Pausar'],
  'transport.reset': ['Back to the first interval', 'Volver al primer intervalo'],
  'transport.speed': ['Playback speed', 'Velocidad de reproducción'],
  'transport.computing': ['Computing the run…', 'Calculando la ejecución…'],

  // ----- the ribbon legend, and what each mark proves -----
  'legend.faultOnset': ['fault onset', 'inicio del fallo'],
  'legend.benignTransient': ['benign transient', 'transitorio benigno'],
  'legend.engineNames': ['engine names it', 'el motor lo identifica'],
  'legend.faultOnset.tip': [
    'Ground truth: the interval a real fault actually began. The engine never sees these ' +
      'marks — it reads only the magnitude trace behind them.',
    'Verdad de referencia: el intervalo en que empezó realmente un fallo. El motor nunca ve ' +
      'estas marcas: solo lee la traza de magnitud que hay detrás.',
  ],
  'legend.benignTransient.tip': [
    'Ground truth: a short, benign disturbance — a prime-time load spike, a one-off home ' +
      'glitch, an isolated reboot. It is not a fault, and the engine must not fire on it. ' +
      'No green mark ever appears beside one; that is what the false-positive rate is counting.',
    'Verdad de referencia: una perturbación breve y benigna — un pico de carga en hora punta, ' +
      'un fallo puntual en un hogar, un reinicio aislado. No es un fallo y el motor no debe ' +
      'dispararse. Nunca aparece una marca verde junto a una: eso es lo que mide la tasa de ' +
      'falsos positivos.',
  ],
  'legend.engineNames.tip': [
    'The engine speaking: the interval it committed to naming the element responsible. It ' +
      'lands after the red mark by design, because evidence has to accumulate — but still ' +
      'before the fault would surface to a customer. That second gap is the lead time.',
    'El motor hablando: el intervalo en que se comprometió a nombrar el elemento responsable. ' +
      'Llega después de la marca roja por diseño, porque la evidencia debe acumularse, pero ' +
      'aun así antes de que el fallo aflorara a un cliente. Esa segunda distancia es la antelación.',
  ],

  // ----- the individual marks on the ribbon -----
  'mark.faultOnset': ['Fault onset (ground truth)', 'Inicio del fallo (verdad de referencia)'],
  'mark.faultOnset.body': [
    'truly begins at interval',
    'empieza realmente en el intervalo',
  ],
  'mark.faultOnset.tail': [
    'The engine is never told — it has to find this in the four-field trace alone.',
    'Al motor nunca se le dice: tiene que encontrarlo solo en la traza de cuatro campos.',
  ],
  'mark.benign': ['Benign transient (ground truth)', 'Transitorio benigno (verdad de referencia)'],
  'mark.benign.body': ['at interval', 'en el intervalo'],
  'mark.benign.tail': [
    'Not a fault. The engine must stay silent here — and no green mark ever follows one.',
    'No es un fallo. El motor debe permanecer en silencio aquí, y nunca le sigue una marca verde.',
  ],
  'mark.named': ['The engine speaking: named', 'El motor hablando: identificó'],
  'mark.named.body': ['at interval', 'en el intervalo'],
  'mark.named.tail': [
    'intervals after the fault began — the wait is evidence accumulating, and it still ' +
      'lands before the fault would surface to a customer.',
    'intervalos después de que empezara el fallo: la espera es evidencia acumulándose, y aun ' +
      'así llega antes de que el fallo aflorara a un cliente.',
  ],

  'measured.boxSwap': ['Box-swap discrimination', 'Discriminación de sustitución del set-top box'],
  'measured.boxSwap.body': [
    'The signature survives a set-top-box replacement, so the box is exonerated and the futile swap is avoided.',
    'La firma sobrevive a la sustitución del set-top box, de modo que el equipo queda exonerado y se evita la sustitución inútil.',
  ],

  // ----- the watching state (before anything is named) -----
  'watch.formingTitle': [
    'A shape is forming — the engine has not yet resolved it',
    'Se está formando una firma: el motor todavía no la ha resuelto',
  ],
  'watch.title': ['Watching the network', 'Vigilando la red'],
  'watch.records': ['four-field records this interval', 'registros de cuatro campos en este intervalo'],
  'watch.meanMagnitude': ['mean magnitude', 'magnitud media'],
  'watch.peak': ['peak', 'pico'],
  'watch.allowedToSee': ['What the engine is allowed to see', 'Lo que el motor puede ver'],
  'watch.allowedToSee.body': [
    'Four fields per record — no error codes, no alarms, no labels. It learns each edge\'s own baseline from the history it observes, and watches the graph for structure.',
    'Cuatro campos por registro: sin códigos de error, sin alarmas, sin etiquetas. Aprende la línea base de cada enlace a partir del histórico que observa y vigila la estructura del grafo.',
  ],
  'watch.drifting': [
    'Something is drifting above baseline right now, but it has not cleared the dwell gate — so nothing is claimed.',
    'Algo se está desviando por encima de la línea base ahora mismo, pero no ha superado el umbral de permanencia, así que no se afirma nada.',
  ],
  'watch.nothingYet': ['Nothing has departed from baseline yet.', 'Todavía nada se ha desviado de la línea base.'],
  'watch.nextVerdict': ['The next verdict lands at interval', 'El siguiente veredicto llega en el intervalo'],
  'watch.holdingFire': ['Holding fire', 'Sin disparar'],
  'watch.decoysFired': ['Decoys fired', 'Señuelos disparados'],
  'watch.falsePositives': ['False positives', 'Falsos positivos'],
  'watch.holdingFire.body': [
    'Benign disturbances — a prime-time surge, a one-off glitch, a reboot — are injected deliberately. An engine that fired on them would be useless in an operations room, so what it ignores is measured just as carefully as what it catches.',
    'Las perturbaciones benignas — un pico en hora punta, un fallo puntual, un reinicio — se inyectan a propósito. Un motor que disparase con ellas sería inútil en una sala de operaciones, así que lo que ignora se mide con el mismo cuidado que lo que detecta.',
  ],
  'watch.formingNotNamed': ['Forming, not yet named', 'Formándose, aún sin identificar'],
  'watch.nothingForming': ['Nothing is forming at this interval.', 'No se está formando nada en este intervalo.'],
  'watch.beganAt': ['began at interval', 'empezó en el intervalo'],
  'watch.namedAt': ['named at', 'identificado en'],
  'receipt.emptyTitle': ['Certified-decision receipt', 'Recibo de decisión certificada'],
  'receipt.empty': [
    'A receipt is written when — and only when — the engine names an element. Until then there is nothing to certify, and the panel stays empty rather than filling with a guess.',
    'Se escribe un recibo cuando —y solo cuando— el motor identifica un elemento. Hasta entonces no hay nada que certificar, y el panel permanece vacío en lugar de llenarse con una suposición.',
  ],

  // ----- the recommended-action list -----
  'action.actOnElement': ['Act on the named element', 'Actuar sobre el elemento identificado'],
  'action.openTicket': ['Open a field ticket', 'Abrir un parte de campo'],
  'action.openTicket.sub': [
    'Pre-filled with the element, its layer, and the receipt',
    'Precargado con el elemento, su capa y el recibo',
  ],
  'action.preWarn': ['Pre-warn Customer Care', 'Avisar a Atención al Cliente'],
  'action.preWarn.sub': ['Suppress the inbound wave from', 'Contener la oleada de llamadas de'],
  'action.shareNoc': ['Share the receipt with the NOC', 'Compartir el recibo con el NOC'],
  'action.shareNoc.sub': [
    'Route the structural evidence to the transport desk',
    'Enviar la evidencia estructural a la mesa de transporte',
  ],
  'receipt.none': ['This verdict carried no receipt.', 'Este veredicto no traía recibo.'],

  // ----- management and map tabs -----
  'map.legend.baseline': ['at baseline', 'en la línea base'],
  'map.noVerdict': [
    'was ever named here. This region stayed at baseline for the whole run.',
    'se identificó aquí. Esta región permaneció en la línea base durante toda la ejecución.',
  ],
  'map.legend.forming': ['shape forming, not yet named', 'firma formándose, aún sin identificar'],


  // ----- management view -----
  'mgmt.pageTitle': ['Structural Intelligence — run scorecard', 'Inteligencia Estructural — marcador de la ejecución'],
  'mgmt.seed': ['seed', 'semilla'],
  'mgmt.computedIn': ['computed in', 'calculado en'],
  'mgmt.againstTruth': [
    'measured against ground truth the engine never sees.',
    'medido frente a una verdad de referencia que el motor nunca ve.',
  ],
  'mgmt.faultsDetected': ['Faults detected', 'Fallos detectados'],
  'mgmt.faultsDetected.d': ['every injected fault, named', 'todos los fallos inyectados, identificados'],
  'mgmt.meanLead': ['Mean lead time', 'Antelación media'],
  'mgmt.meanLead.d': ['before the fault would surface', 'antes de que el fallo aflorara'],
  'mgmt.localization': ['Localization', 'Localización'],
  'mgmt.localization.d': ['the right element, by name', 'el elemento correcto, por su nombre'],
  'mgmt.falsePositives': ['False positives', 'Falsos positivos'],
  'mgmt.spuriousOf': ['spurious of', 'espurios de'],
  'mgmt.quietIntervals': ['quiet intervals', 'intervalos en calma'],
  'mgmt.decoysHeld': ['Decoys held', 'Señuelos contenidos'],
  'mgmt.decoysHeld.d': ['benign spikes that did not fire', 'picos benignos que no dispararon'],
  'mgmt.selfTest': ['Self-test', 'Autoprueba'],
  'mgmt.selfTest.d': ['detected, attributed, box swap avoided', 'detectado, atribuido, sustitución evitada'],
  'mgmt.col.useCase': ['Use case', 'Caso de uso'],
  'mgmt.col.element': ['Element named', 'Elemento identificado'],
  'mgmt.col.detected': ['Detected', 'Detectado'],
  'mgmt.col.lead': ['Lead', 'Antelación'],
  'mgmt.col.localization': ['Localization', 'Localización'],
  'mgmt.col.layer': ['Layer', 'Capa'],
  'mgmt.col.boxSwap': ['Box swap', 'Sustitución STB'],
  'mgmt.decoysFired': ['Decoys fired', 'Señuelos disparados'],
  'mgmt.abstentions': ['Abstentions', 'Abstenciones'],
  'mgmt.abstention.some': [
    'Where the evidence would not resolve to a layer, the engine declined to name one. An abstention is a feature: it is the boundary of its own competence, stated out loud.',
    'Cuando la evidencia no permitía atribuir una capa, el motor se abstuvo de nombrarla. La abstención es una virtud: es el límite de su propia competencia, dicho en voz alta.',
  ],
  'mgmt.abstention.none': [
    'The engine resolved every disturbance it saw in this run.',
    'El motor resolvió todas las perturbaciones que vio en esta ejecución.',
  ],

  // ----- map view -----
  'map.regions': ['Regions in this run', 'Regiones de esta ejecución'],
  'map.verdictsHere': ['Verdicts here', 'Veredictos aquí'],
  'map.networkWeight': ['Network weight', 'Peso en la red'],

  // ----- live deck -----
  'live.watching': [
    'Watching the network live — faults appear the moment the engine names them',
    'Vigilando la red en directo: los fallos aparecen en cuanto el motor los identifica',
  ],

  // ----- states -----
  'state.startingTiny': ['Running the tiny network', 'Ejecutando la red reducida'],
  'state.startingFull': ['Running the full six-region network', 'Ejecutando la red completa de seis regiones'],
  'state.startingBody': [
    'Building the service graph, injecting the faults and the decoys, and letting the engine watch the four-field stream.',
    'Construyendo el grafo de servicio, inyectando los fallos y los señuelos, y dejando que el motor vigile el flujo de cuatro campos.',
  ],
  'state.failed': ['The run did not start', 'La ejecución no ha arrancado'],
  'state.failed.generic': ['Something went wrong.', 'Algo ha salido mal.'],
  'state.failed.needsApi': ['The console needs the API. Start it with', 'La consola necesita la API. Arráncala con'],
  'state.retry': ['Try again', 'Reintentar'],
} as const

export type StringKey = keyof typeof STRINGS

/** A translator bound to one language. */
export type T = (key: StringKey) => string

export function translator(lang: Lang): T {
  const index = lang === 'es' ? 1 : 0
  return (key) => STRINGS[key][index] ?? STRINGS[key][0]
}

/**
 * The layer and shape of a verdict, as words.
 *
 * These read off the engine's own codes, so the label can change language while the
 * value it names cannot. `format.ts` keeps the colours; the wording lives here.
 */
export function layerLabelOf(layer: string | null, t: T): string {
  if (!layer) return t('layerLabel.unresolved')
  const key = `layerLabel.${layer}` as StringKey
  return key in STRINGS ? t(key) : layer
}

export function shapeLabelOf(shape: string, t: T): string {
  const key = `shapeLabel.${shape}` as StringKey
  return key in STRINGS ? t(key) : shape
}

File: backend/si_api/i18n.py
"""
Spanish, and choosing between languages.

WHY SPANISH LIVES HERE AND NOT IN si_core

Which language an operator reads is a property of the request, not of the network. The
domain decides what is true and hands over a Msg; this layer decides which words the
caller gets, exactly as it already decides which JSON shape they get. si_core keeps the
canonical English because the text console and the HTML export have to render with no
API in sight - but nothing beyond English belongs in the domain.

TERMINOLOGY

Equipment and protocol names stay as an operations team writes them: OLT, STB, FEC,
SNR, WAN, CDN, set-top box, gateway, encoder, core. The prose around them is Spanish.
Translating the acronyms would read as unnatural to the engineers this console is for,
and would break the correspondence with the entity ids on the same screen.

A run's numbers never pass through here. Only its sentences do.
"""

from __future__ import annotations

from typing import Dict, Mapping

from si_core.messages import CATALOGUE_EN

DEFAULT_LANG = "en"
SUPPORTED_LANGS = ("en", "es")


CATALOGUE_ES: Dict[str, str] = {
    # ----- what the engine recommends -----
    "action.inspect_access_node":
        "inspeccionar el nodo de acceso y su agregación antes de que escalen las "
        "reclamaciones",
    "action.contact_customer_gateway":
        "contacto proactivo con el cliente y reconfiguración del gateway, sin "
        "desplazamiento de técnico",
    "action.investigate_core_route":
        "investigar la ruta de transporte de core que da servicio a los hogares afectados",
    "action.investigate_content_source":
        "investigar la fuente de contenido y su ruta de entrega",
    "action.watch_and_confirm":
        "puede estar formándose un clúster en este nodo de acceso, vigilar y confirmar",
    "action.escalate_to_human":
        "evidencia insuficiente o ambigua, escalar a revisión humana",
    "action.swap_set_top_box":
        "sustituir el set-top box",

    # ----- the operator headline -----
    "headline.cluster":
        "Nodo de acceso {entity} ({place}) degradándose: {n} hogares con congelaciones "
        "y resintonizaciones en aumento durante los últimos ~{minutes} min, firma de "
        "clúster de acceso local. Recomendación: {action}.",
    "headline.single":
        "Hogar {entity} ({place}) degradándose de forma aislada: congelaciones en "
        "aumento con conectividad doméstica débil durante los últimos ~{minutes} min, "
        "los hogares vecinos del nodo de acceso están sanos, firma de gateway doméstico "
        "aislado. Recomendación: {action}.",
    "headline.path":
        "Ruta de core {entity} degradándose: {n} hogares repartidos en {n_access} nodos "
        "de acceso afectados conjuntamente durante los últimos ~{minutes} min, firma de "
        "transporte de core compartido. Recomendación: {action}.",
    "headline.source":
        "Fuente de contenido {entity} ({channel}) degradándose: {n} hogares repartidos "
        "en {n_access} nodos de acceso que la consumen, afectados durante los últimos "
        "~{minutes} min, firma de fuente de contenido compartida. Recomendación: {action}.",
    "headline.abstention":
        "Límite de competencia: se está formando una perturbación pero la evidencia "
        "todavía no permite atribuirla a una capa. {action_cap}.",
    "headline.all_clear":
        "Todo correcto: ningún fallo en formación en la red monitorizada.",

    # ----- the same headlines as a title -----
    "headline.short.cluster": "Nodo de acceso {entity} degradándose",
    "headline.short.single": "Hogar {entity} en declive",
    "headline.short.path": "Ruta de core {entity} degradándose",
    "headline.short.source": "Fuente de contenido {entity} degradándose",
    "headline.short.abstention": "Límite de competencia",
    "headline.short.all_clear": "Todo correcto",

    # ----- the Structural Health Index bands -----
    "band.degraded_household": "degradado para el hogar",
    "band.critical_element": "crítico para el elemento afectado",
    "band.degraded_element": "degradado para el elemento afectado",
    "band.watch": "en observación",
    "band.healthy": "sano",

    "band.short.degraded_household": "Degradado",
    "band.short.critical_element": "Crítico",
    "band.short.degraded_element": "Degradado",
    "band.short.watch": "En observación",
    "band.short.healthy": "Sano",

    # ----- the certified-decision receipt -----
    "receipt.claim.cluster":
        "Nodo de acceso {entity} degradándose: {n} hogares por detrás con deterioro "
        "creciente",
    "receipt.claim.single":
        "Hogar {entity} degradándose de forma aislada; los hogares vecinos de su nodo "
        "de acceso están sanos",
    "receipt.claim.path":
        "Ruta de core {entity} degradándose: hogares de varios nodos de acceso "
        "afectados conjuntamente",
    "receipt.claim.source":
        "Fuente de contenido {entity} degradándose: hogares sin relación entre sí "
        "afectados conjuntamente",
    "receipt.claim.abstain":
        "Se está formando una perturbación pero la evidencia todavía no permite "
        "atribuirla a una capa",

    "receipt.check.cluster":
        "los hogares afectados comparten un mismo nodo de acceso y sus demás "
        "subsistemas están dentro de la línea base, lo que resulta incompatible con un "
        "fallo de contenido o de core",
    "receipt.check.single":
        "solo este hogar está afectado mientras que los vecinos de su nodo permanecen "
        "en la línea base, lo que resulta incompatible con un fallo de red compartido",
    "receipt.check.path":
        "los hogares afectados abarcan varios nodos de acceso pero comparten una misma "
        "ruta de core, lo que resulta incompatible con un fallo de nodo único o de "
        "contenido",
    "receipt.check.source":
        "los hogares afectados abarcan muchos nodos de acceso pero comparten una misma "
        "fuente de contenido, lo que resulta incompatible con un fallo de nodo de acceso",
    "receipt.check.abstain":
        "la firma estructural todavía no es lo bastante fuerte ni nítida para atribuirla",

    "receipt.support.elevated_sustained":
        "{n} hogares elevados y sostenidos por encima del umbral de permanencia",
    "receipt.support.elevated_sustained.one":
        "1 hogar elevado y sostenido por encima del umbral de permanencia",
    "receipt.support.mean_departure":
        "desviación media de {sigma} desviaciones típicas por encima de la línea base "
        "aprendida",
    "receipt.support.behind_node":
        "{n_behind} de {n} hogares elevados cuelgan de {dst} según el mapa de red",
    "receipt.support.behind_node.one":
        "{n_behind} de 1 hogar elevado cuelga de {dst} según el mapa de red",
    "receipt.support.elevated_near":
        "{n} hogares elevados cerca de {dst}",
    "receipt.support.elevated_near.one":
        "1 hogar elevado cerca de {dst}",
    "receipt.support.convergence_weak":
        "convergencia débil o ambigua; confianza por debajo del umbral de actuación",

    # ----- corroboration joined from enrichment (context, never the basis) -----
    "corrob.access_node":
        "utilización de puertos del nodo de acceso {util:.0%}, errores FEC {fec} "
        "(contexto corroborativo)",
    "corrob.home":
        "SNR del Wi-Fi doméstico {snr} dB, pérdida de paquetes WAN {loss}% "
        "(contexto corroborativo)",
    "corrob.content":
        "fallos de segmento de contenido {segments}, fotogramas descartados por el "
        "encoder {frames} (contexto corroborativo)",
    "corrob.core":
        "carga del transporte de core {load:.0%}, latencia de ruta {latency} ms "
        "(contexto corroborativo)",
    "corrob.example_home":
        "hogar de ejemplo {home}, congelación de {freeze}s en este intervalo",

    # ----- provenance -----
    "provenance.four_field_only":
        "solo el flujo de magnitud de cuatro campos; líneas base aprendidas del histórico",
    "provenance.competence_boundary":
        "límite de competencia alcanzado; se defiere a revisión humana",
    "provenance.derived":
        "derivado del flujo de magnitud de cuatro campos entre los intervalos {start} y "
        "{end}; {note}{signature}",
    "provenance.signature_spans":
        "; la firma abarca {n} hogares (p. ej., {sample})",
    "provenance.signature_spans.one":
        "; la firma abarca 1 hogar ({sample})",

    # ----- the console frame -----
    "console.subtitle":
        "con tecnología de Inteligencia Estructural  |  demo sintética, sin datos de Telefónica",
    "console.config_summary":
        "{homes} hogares, {olts} nodos de acceso, {regions} regiones, {intervals} "
        "intervalos con cadencia de {cadence} min",

    # ----- what each use case is here to show -----
    "panel.title.uc1_network_node":
        "Un nodo de red degradándose (fallo de acceso gradual)",
    "panel.title.uc2_individual":
        "Un hogar concreto degradándose (aislado, los vecinos sanos)",
    "panel.title.uc3_invisible":
        "Un fallo invisible que sobrevive a la sustitución del set-top box",
    "panel.title.uc4_core_path":
        "Una ruta de transporte de core degradándose (un camino entre nodos de acceso)",
    "panel.title.uc5_content_source":
        "Una fuente de contenido degradándose (hogares sin relación, una sola fuente)",

    # ----- the layers, as words in a sentence -----
    "layer.home": "de hogar",
    "layer.access": "de acceso",
    "layer.core": "de core",
    "layer.content": "de contenido",
    "layer.unknown": "desconocida",

    # ----- the four beats -----
    "beat.stream": "Flujo",
    "beat.form": "Formación",
    "beat.predict": "Predicción",
    "beat.prescribe": "Prescripción",
    "beat.stream.text":
        "{homes} hogares de {regions} regiones emiten telemetría de cuatro campos; el "
        "motor aprende lo normal y vigila la estructura del grafo.",
    "beat.form.text.cluster":
        "Se forma un clúster en la capa {layer}: firma de clúster en {entity}.",
    "beat.form.text.single":
        "Se forma un hogar aislado en la capa {layer}: firma individual en {entity}.",
    "beat.form.text.path":
        "Se forma un camino en la capa {layer}: firma de camino en {entity}.",
    "beat.form.text.source":
        "Se forma una fuente en la capa {layer}: firma de fuente en {entity}.",
    "beat.form.text.none":
        "Se forma una firma en la capa {layer}, en {entity}.",
    "beat.predict.text.widening_horizon":
        "Se proyecta que se amplíe ({detail}); con la tendencia actual se extendería "
        "hacia el intervalo {horizon} si no se actúa.",
    "beat.predict.text.widening": "Se proyecta que se amplíe ({detail}).",
    "beat.predict.text.steady":
        "Se proyecta estable; el motor sigue vigilando por si cambia.",
    "beat.prescribe.text":
        "{action_cap}. Señalado {minutes} min antes de que el fallo aflorara; véase el "
        "recibo de decisión certificada.",
    "beat.prescribe.text.no_lead":
        "{action_cap}. Véase el recibo de decisión certificada.",

    # ----- the scorecard, rendered as plain text -----
    "scorecard.heading":
        "Marcador de la demo de Inteligencia Estructural (valores medidos, sin umbrales "
        "preestablecidos)",
    "scorecard.detected": "detectados: {n}/{total}",
    "scorecard.lead_time":
        "antelación de detección: {intervals} intervalos ({minutes}) antes de que el "
        "fallo aflorara",
    "scorecard.localization": "precisión de localización: {pct}",
    "scorecard.layer_attribution": "atribución de capa: {pct}",
    "scorecard.box_swap":
        "discriminación de sustitución de set-top box: {pct} (se habría evitado la "
        "sustitución)",
    "scorecard.action_correctness": "corrección de la acción: {pct}",
    "scorecard.note": "nota: {note}",
    "scorecard.false_positive_rate":
        "tasa de falsos positivos: {pct} ({spurious} espurios de {total} intervalos sin fallo)",
    "scorecard.decoys_fired": "señuelos que dispararon: {fired}/{total}",
    "score.note.not_detected": "no detectado",
    "score.note.detected":
        "detectado en el intervalo {interval}, identificado {entity} ({layer}/{shape}), "
        "{lead} intervalos de antelación respecto al afloramiento en {surfacing}",

    # ----- the forward projection -----
    "trajectory.widening": "convergencia ampliándose",
    "trajectory.steady": "convergencia estable",
}


CATALOGUES: Dict[str, Mapping[str, str]] = {
    "en": CATALOGUE_EN,
    "es": CATALOGUE_ES,
}


def normalise_lang(value: str | None) -> str:
    """The language the caller asked for, reduced to one this demo actually ships.

    Accepts what a browser sends as well as what the UI sends: "es", "es-ES",
    "es-ES,es;q=0.9,en;q=0.8". Anything unrecognised falls back to English rather than
    failing the request - a console in the wrong language is recoverable, a 400 in the
    middle of a demo is not.
    """
    if not value:
        return DEFAULT_LANG
    for part in value.split(","):
        tag = part.split(";")[0].strip().lower()
        if not tag:
            continue
        primary = tag.split("-")[0]
        if primary in CATALOGUES:
            return primary
    return DEFAULT_LANG


def catalogue_for(lang: str | None) -> Mapping[str, str]:
    return CATALOGUES[normalise_lang(lang)]
